src/main/scala/is/hail/methods/Aggregators.scala

```diff
@@ -495,3 +492,47 @@ class TakeByAggregator[T](var t: Type, var f: (Any) => Any, var n: Int)(implicit
 
   def copy() = new TakeByAggregator(t, f, n)
 }
+
+class KeyedAggregator[T, K](aggregator: TypedAggregator[T]) extends TypedAggregator[Map[Any, T]] {
+
+  private val m = mutable.Map[Any, TypedAggregator[T]]()
+
+  def result = m.map { case (k, v) => (k, v.result) }.toMap
+
+  def seqOp(x: Any) {
+    val cx = x.asInstanceOf[Row]
+    if (cx != null)
+      seqOp(cx.get(0), cx.get(1))
+    else
+      seqOp(null, null)
+  }
+
+  private def seqOp(key: Any, x: Any) {
+    val agg = m.getOrElseUpdate(key, aggregator.copy())
```

Comment:
I think for an aggregator's seqop, which might get called once for every genotype, we should avoid a method that takes a lazy argument _and_ allocates an option by its internal call to `get`. Sadly, I'm not aware of any way to do this efficiently in Scala. I think we should use `java.util.HashMap`. You can call `m.get` which returns `null` if the key doesn't exist. so it can be:
```scala
var agg = m.get(key)
if (agg == null) {
  agg = aggregator.copy()
  m.put(key, agg)
}
```